File: mediation/mediatedviews/GoogleAdMob/ANAdAdapterInterstitialAdMob.m
/*   Copyright 2013 APPNEXUS INC
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "ANAdAdapterInterstitialAdMob.h"
#import "ANAdAdapterBaseDFP.h"

@interface ANAdAdapterInterstitialAdMob ()

//@property (nonatomic, readwrite, strong) GADInterstitial *interstitialAd;
@property(nonatomic, strong) GADInterstitialAd *interstitialAd;

@end

@implementation ANAdAdapterInterstitialAdMob
@synthesize delegate;

#pragma mark ANCustomAdapterInterstitial

- (void)requestInterstitialAdWithParameter:(nullable NSString *)parameterString
                                  adUnitId:(nullable NSString *)idString
                       targetingParameters:(nullable ANTargetingParameters *)targetingParameters
{
    ANLogDebug(@"Requesting AdMob interstitial");
    [GADInterstitialAd loadWithAdUnitID:idString
                                request:[self createRequestFromTargetingParameters:targetingParameters]
                      completionHandler:^(GADInterstitialAd *ad, NSError *error) {
      if (error) {
        ANLogError(@"Failed to load interstitial ad with error: %@", [error localizedDescription]);
          [self.delegate didFailToLoadAd:[ANAdAdapterBaseDFP responseCodeFromRequestError:error]];
          return;
      }
      ANLogDebug(@"AdMob interstitial did load");
      [self.delegate didLoadInterstitialAd:self];
      self.interstitialAd = ad;
      self.interstitialAd.fullScreenContentDelegate = self;
    }];
}

- (void)presentFromViewController:(UIViewController *)viewController
{
    if (self.interstitialAd && [self.interstitialAd
                                canPresentFromRootViewController:viewController
                                error:nil]) {
        ANLogDebug(@"Showing AdMob interstitial");
        [self.interstitialAd presentFromRootViewController:viewController];
    } else {
        ANLogDebug(@"AdMob interstitial was unavailable");
        [self.delegate failedToDisplayAd];
        return;
    }
}

- (GADRequest *)createRequestFromTargetingParameters:(ANTargetingParameters *)targetingParameters {
    return [ANAdAdapterBaseDFP googleAdMobRequestFromTargetingParameters:targetingParameters rootViewController:nil];
}

#pragma mark GADFullScreenContentDelegate

- (void)ad:(nonnull id<GADFullScreenPresentingAd>)ad didFailToPresentFullScreenContentWithError:(nonnull NSError *)error {
    [self.delegate failedToDisplayAd];
}

- (void)adWilllPresentFullScreenContent:(nonnull id<GADFullScreenPresentingAd>)ad {
    [self.delegate willPresentAd];
}

-(void) adWillDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad{
    ANLogDebug(@"AdMob interstitial will close");
    [self.delegate willCloseAd];
}

- (void)adDidDismissFullScreenContent:(nonnull id<GADFullScreenPresentingAd>)ad {
    ANLogDebug(@"AdMob interstitial did close");
    [self.delegate didCloseAd];
}

- (void)adDidRecordImpression:(nonnull id<GADFullScreenPresentingAd>)ad{
    ANLogDebug(@"AdMob interstitial impression recorded");
    [self.delegate adDidLogImpression];
}

// Tells the delegate that a click has been recorded for the ad.
- (void)adDidRecordClick:(nonnull id<GADFullScreenPresentingAd>)ad{
    [self.delegate adWasClicked];
}

@end

File: examples/SimpleMediation/SimpleMediation/ANAdAdapterInterstitialAdMob.swift
//
//  CustomAdapter.swift
//  SimpleMediation
//
//  Created by System on 17/05/22.
//  Copyright © 2022 Xandr. All rights reserved.
//

import Foundation
import AppNexusSDK
import GoogleMobileAds

@objc(ANAdAdapterInterstitialAdMob)
public class ANAdAdapterInterstitialAdMob : NSObject , ANCustomAdapterInterstitial , GADFullScreenContentDelegate {
    
    
    public var delegate: AnyObject?
    private var interstitial: GADInterstitialAd?
    
    
    public func requestAd(withParameter parameterString: String?, adUnitId idString: String?, targetingParameters: ANTargetingParameters?) {
        let request = GADRequest()
        GADInterstitialAd.load(withAdUnitID:"ca-app-pub-3940256099942544/4411468910",
                               request: request,
                               completionHandler: { [self] ad, error in
            if let error = error {
                print("Failed to load interstitial ad with error: \(error.localizedDescription)")
                return
            }
            
            interstitial = ad
            
            if interstitial != nil {
                self.delegate?.didLoadInterstitialAd(self)
                interstitial?.fullScreenContentDelegate = self;
                
            } else {
                self.delegate?.didFail(toLoadAd: ANAdResponseCode.unable_TO_FILL())
                print("Ad wasn't ready")
            }
        }
        )
    }
    
    public func present(from viewController: UIViewController?) {
        //        if (interstitial != nil) && try! interstitial?.canPresent(fromRootViewController: viewController!) != nil {
        interstitial!.present(fromRootViewController: viewController!)
        //        }
    }
    
    
    /// Tells the delegate that the ad failed to present full screen content.
    public func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        print("Ad did fail to present full screen content.")
        self.delegate?.failedToDisplayAd()
    }
    
    /// Tells the delegate that the ad will present full screen content.
    public func adWillPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        print("Ad will present full screen content.")
        self.delegate?.willPresentAd()
    }
    
    /// Tells the delegate that the ad dismissed full screen content.
    public func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        print("Ad did dismiss full screen content.")
        self.delegate?.didCloseAd()
    }
    
    public func adWillDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        self.delegate?.willCloseAd()
    }
    public func adDidRecordClick(_ ad: GADFullScreenPresentingAd) {
        self.delegate?.adWasClicked()
        
    }
    
    public func adDidRecordImpression(_ ad: GADFullScreenPresentingAd) {
        self.delegate?.adDidLogImpression()
        
    }
}
